main.tf

```diff
@@ -1,25 +1,46 @@
 data "aws_region" "current" {}
 
+# Seach for ubuntu AMI based on supplied AMI parameters
 data "aws_ami" "ubuntu" {
   most_recent = true
 
   filter {
     name   = "name"
-    values = ["ubuntu/images/hvm-ssd/ubuntu-focal-20.04-amd64-server-*"]
+    values = [var.ami_name_filter]
   }
 
   filter {
     name   = "virtualization-type"
     values = ["hvm"]
   }
 
-  owners = ["099720109477"] # Canonical
+  owners = [var.ami_owner]
+}
+
+# Find encryption status of located image
+locals {
+  block_device_mappings = {
+    for device in data.aws_ami.ubuntu.block_device_mappings : device.device_name => device
+  }
+}
+
+# If the user has selected to do a copy-down of the AMI to their local account, do so and use it as the default.
+# This prevents AMI lifecycle events (disappearing AMIs for instance) from affecting ASG
+resource "aws_ami_copy" "tfe_copy" {
+  count             = var.ami_copy ? 1 : 0
+  name              = "${var.friendly_name_prefix}-${data.aws_ami.ubuntu.name}-local-copy"
+  source_ami_id     = data.aws_ami.ubuntu.id
+  source_ami_region = data.aws_region.current.name
+  encrypted         = local.block_device_mappings["/dev/sda1"].ebs.encrypted
+  tags = {
+    Name = "${var.friendly_name_prefix}-${data.aws_ami.ubuntu.name}-local-copy"
+  }
```

Comment:
While I can do this, is this of genuine benefit to the module complexity? This will mean either having to hard-code a specific AMI ID in the variables or have a required variable which messes with the module flow.

In my own modules typically I have something like what is posted above (allowing the user to do a custom data filter.

Reply:
> While I can do this, is this of genuine benefit to the module complexity? This will mean either having to hard-code a specific AMI ID in the variables or have a required variable which messes with the module flow.

I appreciate your concern here, but as is evidenced by the proposed changes, having this sort of data source embedded in the module encourages interface sprawl as we try to accommodate multiple use cases rather than letting the caller locate and provide an AMI in any manner that they prefer.

Reply:
I've replaced the data object with ami_id alone and extended the health check timeouts to a variable option.